src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs

```diff
@@ -72,21 +73,44 @@ private static unsafe IPHostEntry CreateIPHostEntry(Interop.Sys.HostEntry hostEn
             }
             else
             {
-                ipAddresses = new IPAddress[numAddresses];
-
-                void* addressListHandle = hostEntry.AddressListHandle;
-                var nativeIPAddress = default(Interop.Sys.IPAddress);
-                for (int i = 0; i < numAddresses; i++)
+                //
+                // getaddrinfo returns multiple entries per address, for each socket type (datagram, stream, etc.).
+                // Our callers expect just one entry for each address.  So we need to deduplicate the results.
+                // It's important to keep the addresses in order, since they are returned in the order in which
+                // connections should be attempted.
+                // 
+                // We assume that the list returned by getaddrinfo is relatively short; after all, the intent is that
+                // the caller may need to attempt to contact every address in the list before giving up on a connection
+                // attempt.  So an O(N^2) algorithm should be fine here.  Keep in mind that any "better" algorithm 
+                // is likely to involve extra allocations, hashing, etc., and so will probably be more expensive than
+                // this one in the typical (short list) case.
+                //
+                var nativeAddresses = new Interop.Sys.IPAddress[hostEntry.IPAddressCount];
+                var nativeAddressCount = 0;
+
+                var addressListHandle = hostEntry.AddressListHandle;
+                for (int i = 0; i < hostEntry.IPAddressCount; i++)
                 {
+                    var nativeIPAddress = default(Interop.Sys.IPAddress);
                     int err = Interop.Sys.GetNextIPAddress(&hostEntry, &addressListHandle, &nativeIPAddress);
                     Debug.Assert(err == 0);
 
-                    ipAddresses[i] = nativeIPAddress.GetIPAddress();
+                    if (Array.IndexOf(nativeAddresses, nativeIPAddress, 0, nativeAddressCount) == -1)
+                    {
+                        nativeAddresses[nativeAddressCount] = nativeIPAddress;
+                        nativeAddressCount++;
+                    }
+                }
+
+                ipAddresses = new IPAddress[nativeAddressCount];
+                for (int i = 0; i < nativeAddressCount; i++)
+                {
+                    ipAddresses[i] = nativeAddresses[i].GetIPAddress();
                 }
             }
 
             int numAliases;
-            for (numAliases = 0; hostEntry.Aliases[numAliases] != null; numAliases++)
+            for (numAliases = 0; hostEntry.Aliases != null && hostEntry.Aliases[numAliases] != null; numAliases++)
```

Comment:
GitHub won't let me comment on it, but below there's a FreeHostEntry call.  Unrelated to your change, should we put that in a finally?  There's a lot of code between the allocation and the free, such that it seems like the potential source of a leak in case something goes wrong.